File: hackAstone_app/hackAstone_app/ArenaBilingualParsing.swift
import Foundation

struct BattleLocaleSlice: Equatable {
    var question: String
    var category: String
    var builderView: String
    var breakerView: String
    var judgeQuestions: [String]
    var reveal: String

    func toBattle(id: String) -> Battle {
        Battle(
            id: id,
            question: question,
            category: category,
            builderView: builderView,
            breakerView: breakerView,
            judgeQuestions: judgeQuestions,
            reveal: reveal
        )
    }
}

struct DisciplineSummaryBilingualParsed: Equatable {
    let en: String
    let zh: String
}

struct DisciplineDualReplyParsed: Equatable {
    let builder: String
    let breaker: String
}

struct DisciplineBattleBilingualParsed: Equatable {
    let en: BattleLocaleSlice
    let zh: BattleLocaleSlice
}

struct DilemmaTurnSliceParsed: Equatable {
    let philosopherReply: String
    let judgeQuestion: String
    let continueDebate: Bool
}

struct DilemmaTurnBilingualParsed: Equatable {
    let en: DilemmaTurnSliceParsed
    let zh: DilemmaTurnSliceParsed

    func pick(english: Bool) -> DilemmaTurnSliceParsed {
        english ? en : zh
    }
}

struct DilemmaSummaryBilingualParsed: Equatable {
    let en: String
    let zh: String

    func pick(english: Bool) -> String {
        english ? en : zh
    }
}

struct RoundtableMessageSlice: Equatable {
    let speaker: String
    let content: String
}

struct RoundtableMessagesBilingualParsed: Equatable {
    let en: [RoundtableMessageSlice]
    let zh: [RoundtableMessageSlice]

    func pick(english: Bool) -> [RoundtableMessageSlice] {
        english ? en : zh
    }
}

enum ArenaBilingualParsing {

    private static func pickLocaleBlock(_ root: [String: Any], keys: [String]) -> [String: Any]? {
        for key in keys {
            if let v = root[key] as? [String: Any] { return v }
        }
        let lower = Set(keys.map { $0.lowercased() })
        for (k, v) in root {
            if lower.contains(k.lowercased()), let o = v as? [String: Any] { return o }
        }
        return nil
    }

    private static func normalizeBattleSlice(_ raw: Any?) -> BattleLocaleSlice? {
        guard let o = raw as? [String: Any] else { return nil }
        let judgeRaw = o["judgeQuestions"] ?? o["judge_questions"]
        let judgeQuestions: [String] = (judgeRaw as? [Any])?
            .map { String(describing: $0).trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty } ?? []
        let slice = BattleLocaleSlice(
            question: String(describing: o["question"] ?? "").trimmingCharacters(in: .whitespacesAndNewlines),
            category: String(describing: o["category"] ?? "General").trimmingCharacters(in: .whitespacesAndNewlines),
            builderView: String(describing: o["builderView"] ?? o["builder_view"] ?? "").trimmingCharacters(in: .whitespacesAndNewlines),
            breakerView: String(describing: o["breakerView"] ?? o["breaker_view"] ?? "").trimmingCharacters(in: .whitespacesAndNewlines),
            judgeQuestions: judgeQuestions,
            reveal: String(describing: o["reveal"] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        )
        guard !slice.question.isEmpty, !slice.builderView.isEmpty, !slice.breakerView.isEmpty else { return nil }
        return slice
    }

    static func parseDisciplineBattle(from text: String, structured: Any?) -> DisciplineBattleBilingualParsed? {
        if let dict = structured as? [String: Any] {
            if let parsed = parseDisciplineBattleDict(dict) { return parsed }
        }
        guard let data = text.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        return parseDisciplineBattleDict(root)
    }

    private static func parseDisciplineBattleDict(_ root: [String: Any]) -> DisciplineBattleBilingualParsed? {
        if let en = normalizeBattleSlice(root["en"]),
           let zh = normalizeBattleSlice(root["zh"]) {
            return DisciplineBattleBilingualParsed(en: en, zh: zh)
        }
        if let en = normalizeBattleSlice(pickLocaleBlock(root, keys: ["en", "english"])),
           let zh = normalizeBattleSlice(pickLocaleBlock(root, keys: ["zh", "chinese", "cn"])) {
            return DisciplineBattleBilingualParsed(en: en, zh: zh)
        }
        if let single = normalizeBattleSlice(root) {
            return DisciplineBattleBilingualParsed(en: single, zh: single)
        }
        return nil
    }

    static func parseDilemmaTurn(from text: String, structured: Any?) -> DilemmaTurnBilingualParsed? {
        if let dict = structured as? [String: Any], let p = parseDilemmaTurnDict(dict) { return p }
        guard let data = text.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        return parseDilemmaTurnDict(root)
    }

    private static func parseDilemmaTurnDict(_ root: [String: Any]) -> DilemmaTurnBilingualParsed? {
        func slice(_ raw: Any?) -> DilemmaTurnSliceParsed? {
            guard let o = raw as? [String: Any] else { return nil }
            let pr = String(describing: o["philosopherReply"] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            let jq = String(describing: o["judgeQuestion"] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            guard !pr.isEmpty, !jq.isEmpty else { return nil }
            let cont = (o["continueDebate"] as? Bool) ?? true
            return DilemmaTurnSliceParsed(philosopherReply: pr, judgeQuestion: jq, continueDebate: cont)
        }
        if let en = slice(root["en"]), let zh = slice(root["zh"]) {
            return DilemmaTurnBilingualParsed(en: en, zh: zh)
        }
        if let en = slice(pickLocaleBlock(root, keys: ["en", "english"])),
           let zh = slice(pickLocaleBlock(root, keys: ["zh", "chinese", "cn"])) {
            return DilemmaTurnBilingualParsed(en: en, zh: zh)
        }
        if let single = slice(root) {
            return DilemmaTurnBilingualParsed(en: single, zh: single)
        }
        return nil
    }

    static func parseDilemmaSummary(from text: String, structured: Any?) -> DilemmaSummaryBilingualParsed? {
        if let dict = structured as? [String: Any], let p = parseDilemmaSummaryDict(dict) { return p }
        guard let data = text.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        return parseDilemmaSummaryDict(root)
    }

    private static func parseDilemmaSummaryDict(_ root: [String: Any]) -> DilemmaSummaryBilingualParsed? {
        func explanation(_ raw: Any?) -> String? {
            guard let o = raw as? [String: Any] else { return nil }
            let fe = String(describing: o["fullExplanation"] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            return fe.isEmpty ? nil : fe
        }
        if let en = explanation(root["en"]), let zh = explanation(root["zh"]) {
            return DilemmaSummaryBilingualParsed(en: en, zh: zh)
        }
        if let en = explanation(pickLocaleBlock(root, keys: ["en", "english"])),
           let zh = explanation(pickLocaleBlock(root, keys: ["zh", "chinese", "cn"])) {
            return DilemmaSummaryBilingualParsed(en: en, zh: zh)
        }
        let fe = String(describing: root["fullExplanation"] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        if !fe.isEmpty {
            return DilemmaSummaryBilingualParsed(en: fe, zh: fe)
        }
        return nil
    }

    static func parseRoundtableMessages(from text: String, structured: Any?) -> RoundtableMessagesBilingualParsed? {
        if let dict = structured as? [String: Any], let p = parseRoundtableDict(dict) { return p }
        guard let data = text.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        return parseRoundtableDict(root)
    }

    private static func parseRoundtableDict(_ root: [String: Any]) -> RoundtableMessagesBilingualParsed? {
        func messages(_ raw: Any?) -> [RoundtableMessageSlice]? {
            guard let o = raw as? [String: Any],
                  let arr = o["messages"] as? [[String: Any]]
            else { return nil }
            let mapped = arr.compactMap { item -> RoundtableMessageSlice? in
                let speaker = String(describing: item["speaker"] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
                let content = String(describing: item["content"] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
                guard !speaker.isEmpty, !content.isEmpty else { return nil }
                return RoundtableMessageSlice(speaker: speaker, content: content)
            }
            return mapped.isEmpty ? nil : mapped
        }
        if let en = messages(root["en"]), let zh = messages(root["zh"]) {
            return RoundtableMessagesBilingualParsed(en: en, zh: zh)
        }
        if let en = messages(pickLocaleBlock(root, keys: ["en", "english"])),
           let zh = messages(pickLocaleBlock(root, keys: ["zh", "chinese", "cn"])) {
            return RoundtableMessagesBilingualParsed(en: en, zh: zh)
        }
        if let single = messages(root) {
            return RoundtableMessagesBilingualParsed(en: single, zh: single)
        }
        return nil
    }

    static func parseDisciplineSummary(from text: String, structured: Any?) -> DisciplineSummaryBilingualParsed? {
        if let dict = structured as? [String: Any], let p = parseDisciplineSummaryDict(dict) { return p }
        guard let data = text.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        return parseDisciplineSummaryDict(root)
    }

    private static func parseDisciplineSummaryDict(_ root: [String: Any]) -> DisciplineSummaryBilingualParsed? {
        func summary(_ raw: Any?) -> String? {
            guard let o = raw as? [String: Any] else { return nil }
            let s = String(describing: o["summary"] ?? o["reveal"] ?? o["fullExplanation"] ?? "")
                .trimmingCharacters(in: .whitespacesAndNewlines)
            return s.isEmpty ? nil : s
        }
        if let en = summary(root["en"]), let zh = summary(root["zh"]) {
            return DisciplineSummaryBilingualParsed(en: en, zh: zh)
        }
        if let en = summary(pickLocaleBlock(root, keys: ["en", "english"])),
           let zh = summary(pickLocaleBlock(root, keys: ["zh", "chinese", "cn"])) {
            return DisciplineSummaryBilingualParsed(en: en, zh: zh)
        }
        let single = summary(root)
        if let single { return DisciplineSummaryBilingualParsed(en: single, zh: single) }
        return nil
    }

    static func parseDisciplineDual(from text: String, structured: Any?) -> DisciplineDualReplyParsed? {
        if let dict = structured as? [String: Any],
           let b = dict["builder"] as? String,
           let k = dict["breaker"] as? String,
           !b.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
           !k.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return DisciplineDualReplyParsed(builder: b.trimmingCharacters(in: .whitespacesAndNewlines),
                                             breaker: k.trimmingCharacters(in: .whitespacesAndNewlines))
        }
        return parseDisciplineDualMarkers(from: finalizeStreamSpeech(text))
    }

    static func finalizeStreamSpeech(_ raw: String) -> String {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.hasPrefix("{"),
              let data = trimmed.data(using: .utf8),
              let o = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let c = o["content"] as? String,
              !c.isEmpty
        else { return trimmed }
        return c
    }

    private static func parseDisciplineDualMarkers(from text: String) -> DisciplineDualReplyParsed? {
        let patternBuilder = #"(?m)^\s*(?:\[Builder\]|【建构者】|建构者[:：])\s*"#
        let patternBreaker = #"(?m)^\s*(?:\[Breaker\]|【破坏者】|破坏者[:：])\s*"#
        guard let bRange = text.range(of: patternBuilder, options: .regularExpression),
              let kRange = text.range(of: patternBreaker, options: .regularExpression),
              bRange.lowerBound < kRange.lowerBound
        else { return nil }
        let builderText = String(text[bRange.upperBound..<kRange.lowerBound]).trimmingCharacters(in: .whitespacesAndNewlines)
        let breakerText = String(text[kRange.upperBound...]).trimmingCharacters(in: .whitespacesAndNewlines)
        guard !builderText.isEmpty, !breakerText.isEmpty else { return nil }
        return DisciplineDualReplyParsed(builder: builderText, breaker: breakerText)
    }

    static func buildDebateNoteKey(philosopherId: String, question: String) -> String {
        "\(philosopherId)|\(question.trimmingCharacters(in: .whitespacesAndNewlines))"
    }
}

File: hackAstone_Hams/entry/.preview/default/cache/default/default@PreviewArkTS/esmodule/debug/entry/src/main/ets/common/ArenaBilingualParsing.ts
import type { Battle } from './ArenaModels';
export interface BattleLocaleSliceInit {
    question?: string;
    category?: string;
    builderView?: string;
    breakerView?: string;
    judgeQuestions?: string[];
    reveal?: string;
}
export class BattleLocaleSlice {
    question: string = '';
    category: string = '';
    builderView: string = '';
    breakerView: string = '';
    judgeQuestions: string[] = [];
    reveal: string = '';
    constructor(init?: BattleLocaleSliceInit) {
        if (init !== undefined) {
            this.question = init.question ?? '';
            this.category = init.category ?? '';
            this.builderView = init.builderView ?? '';
            this.breakerView = init.breakerView ?? '';
            this.judgeQuestions = init.judgeQuestions ?? [];
            this.reveal = init.reveal ?? '';
        }
    }
    toBattle(id: string): Battle {
        const battle: Battle = {
            id: id,
            question: this.question,
            category: this.category,
            builderView: this.builderView,
            breakerView: this.breakerView,
            judgeQuestions: this.judgeQuestions,
            reveal: this.reveal,
        };
        return battle;
    }
}
export class DisciplineSummaryBilingualParsed {
    en: string = '';
    zh: string = '';
    constructor(en: string, zh: string) {
        this.en = en;
        this.zh = zh;
    }
}
export class DisciplineDualReplyParsed {
    builder: string = '';
    breaker: string = '';
    constructor(builder: string, breaker: string) {
        this.builder = builder;
        this.breaker = breaker;
    }
}
export class DisciplineBattleBilingualParsed {
    en: BattleLocaleSlice = new BattleLocaleSlice();
    zh: BattleLocaleSlice = new BattleLocaleSlice();
    constructor(en: BattleLocaleSlice, zh: BattleLocaleSlice) {
        this.en = en;
        this.zh = zh;
    }
}
export class DilemmaTurnSliceParsed {
    philosopherReply: string = '';
    judgeQuestion: string = '';
    continueDebate: boolean = true;
    constructor(philosopherReply: string, judgeQuestion: string, continueDebate: boolean) {
        this.philosopherReply = philosopherReply;
        this.judgeQuestion = judgeQuestion;
        this.continueDebate = continueDebate;
    }
}
export class DilemmaTurnBilingualParsed {
    en: DilemmaTurnSliceParsed = new DilemmaTurnSliceParsed('', '', true);
    zh: DilemmaTurnSliceParsed = new DilemmaTurnSliceParsed('', '', true);
    constructor(en: DilemmaTurnSliceParsed, zh: DilemmaTurnSliceParsed) {
        this.en = en;
        this.zh = zh;
    }
    pick(english: boolean): DilemmaTurnSliceParsed {
        return english ? this.en : this.zh;
    }
}
export class DilemmaSummaryBilingualParsed {
    en: string = '';
    zh: string = '';
    constructor(en: string, zh: string) {
        this.en = en;
        this.zh = zh;
    }
    pick(english: boolean): string {
        return english ? this.en : this.zh;
    }
}
export class RoundtableMessageSlice {
    speaker: string = '';
    content: string = '';
    constructor(speaker: string, content: string) {
        this.speaker = speaker;
        this.content = content;
    }
}
export class RoundtableMessagesBilingualParsed {
    en: RoundtableMessageSlice[] = [];
    zh: RoundtableMessageSlice[] = [];
    constructor(en: RoundtableMessageSlice[], zh: RoundtableMessageSlice[]) {
        this.en = en;
        this.zh = zh;
    }
    pick(english: boolean): RoundtableMessageSlice[] {
        return english ? this.en : this.zh;
    }
}
export class ArenaBilingualParsing {
    private static pickLocaleBlock(root: Record<string, Object>, keys: string[]): Record<string, Object> | undefined {
        for (const key of keys) {
            const v = root[key];
            if (v !== undefined && v !== null && typeof v === 'object' && !Array.isArray(v)) {
                return v as Record<string, Object>;
            }
        }
        const lower = new Set(keys.map((k) => k.toLowerCase()));
        for (const k of Object.keys(root)) {
            if (lower.has(k.toLowerCase())) {
                const v = root[k];
                if (v !== undefined && v !== null && typeof v === 'object' && !Array.isArray(v)) {
                    return v as Record<string, Object>;
                }
            }
        }
        return undefined;
    }
    private static str(v: Object | undefined): string {
        if (v === undefined || v === null) {
            return '';
        }
        return String(v).trim();
    }
    private static normalizeBattleSlice(raw: Object | null | undefined): BattleLocaleSlice | null {
        if (raw === undefined || raw === null || typeof raw !== 'object' || Array.isArray(raw)) {
            return null;
        }
        const o = raw as Record<string, Object>;
        const judgeRaw = o['judgeQuestions'] ?? o['judge_questions'];
        let judgeQuestions: string[] = [];
        if (Array.isArray(judgeRaw)) {
            judgeQuestions = judgeRaw
                .map((item: Object) => String(item).trim())
                .filter((item: string) => item.length > 0);
        }
        const slice = new BattleLocaleSlice({
            question: ArenaBilingualParsing.str(o['question']),
            category: ArenaBilingualParsing.str(o['category'] || 'General'),
            builderView: ArenaBilingualParsing.str(o['builderView'] ?? o['builder_view']),
            breakerView: ArenaBilingualParsing.str(o['breakerView'] ?? o['breaker_view']),
            judgeQuestions,
            reveal: ArenaBilingualParsing.str(o['reveal']),
        });
        if (slice.question.length === 0 || slice.builderView.length === 0 || slice.breakerView.length === 0) {
            return null;
        }
        return slice;
    }
    static parseDisciplineBattle(sourceText: string, structured: Object | null): DisciplineBattleBilingualParsed | null {
        if (structured !== null && typeof structured === 'object' && !Array.isArray(structured)) {
            const parsed = ArenaBilingualParsing.parseDisciplineBattleDict(structured as Record<string, Object>);
            if (parsed !== null) {
                return parsed;
            }
        }
        try {
            const root = JSON.parse(sourceText) as Record<string, Object>;
            return ArenaBilingualParsing.parseDisciplineBattleDict(root);
        }
        catch (_e) {
            return null;
        }
    }
    private static parseDisciplineBattleDict(root: Record<string, Object>): DisciplineBattleBilingualParsed | null {
        const directEn = ArenaBilingualParsing.normalizeBattleSlice(root['en']);
        const directZh = ArenaBilingualParsing.normalizeBattleSlice(root['zh']);
        if (directEn !== null && directZh !== null) {
            return new DisciplineBattleBilingualParsed(directEn, directZh);
        }
        const en = ArenaBilingualParsing.normalizeBattleSlice(ArenaBilingualParsing.pickLocaleBlock(root, ['en', 'english']));
        const zh = ArenaBilingualParsing.normalizeBattleSlice(ArenaBilingualParsing.pickLocaleBlock(root, ['zh', 'chinese', 'cn']));
        if (en !== null && zh !== null) {
            return new DisciplineBattleBilingualParsed(en, zh);
        }
        const single = ArenaBilingualParsing.normalizeBattleSlice(root);
        if (single !== null) {
            return new DisciplineBattleBilingualParsed(single, single);
        }
        return null;
    }
    static parseDilemmaTurn(sourceText: string, structured: Object | null): DilemmaTurnBilingualParsed | null {
        if (structured !== null && typeof structured === 'object' && !Array.isArray(structured)) {
            const p = ArenaBilingualParsing.parseDilemmaTurnDict(structured as Record<string, Object>);
            if (p !== null) {
                return p;
            }
        }
        try {
            const root = JSON.parse(sourceText) as Record<string, Object>;
            return ArenaBilingualParsing.parseDilemmaTurnDict(root);
        }
        catch (_e) {
            return null;
        }
    }
    private static sliceDilemmaTurn(raw: Object | undefined): DilemmaTurnSliceParsed | null {
        if (raw === undefined || raw === null || typeof raw !== 'object' || Array.isArray(raw)) {
            return null;
        }
        const o = raw as Record<string, Object>;
        const pr = ArenaBilingualParsing.str(o['philosopherReply']);
        const jq = ArenaBilingualParsing.str(o['judgeQuestion']);
        if (pr.length === 0 || jq.length === 0) {
            return null;
        }
        const cont = typeof o['continueDebate'] === 'boolean' ? (o['continueDebate'] as boolean) : true;
        return new DilemmaTurnSliceParsed(pr, jq, cont);
    }
    private static parseDilemmaTurnDict(root: Record<string, Object>): DilemmaTurnBilingualParsed | null {
        const directEn = ArenaBilingualParsing.sliceDilemmaTurn(root['en']);
        const directZh = ArenaBilingualParsing.sliceDilemmaTurn(root['zh']);
        if (directEn !== null && directZh !== null) {
            return new DilemmaTurnBilingualParsed(directEn, directZh);
        }
        const en = ArenaBilingualParsing.sliceDilemmaTurn(ArenaBilingualParsing.pickLocaleBlock(root, ['en', 'english']));
        const zh = ArenaBilingualParsing.sliceDilemmaTurn(ArenaBilingualParsing.pickLocaleBlock(root, ['zh', 'chinese', 'cn']));
        if (en !== null && zh !== null) {
            return new DilemmaTurnBilingualParsed(en, zh);
        }
        const single = ArenaBilingualParsing.sliceDilemmaTurn(root);
        if (single !== null) {
            return new DilemmaTurnBilingualParsed(single, single);
        }
        return null;
    }
    static parseDilemmaSummary(sourceText: string, structured: Object | null): DilemmaSummaryBilingualParsed | null {
        if (structured !== null && typeof structured === 'object' && !Array.isArray(structured)) {
            const p = ArenaBilingualParsing.parseDilemmaSummaryDict(structured as Record<string, Object>);
            if (p !== null) {
                return p;
            }
        }
        try {
            const root = JSON.parse(sourceText) as Record<string, Object>;
            return ArenaBilingualParsing.parseDilemmaSummaryDict(root);
        }
        catch (_e) {
            return null;
        }
    }
    private static explanation(raw: Object | undefined): string | null {
        if (raw === undefined || raw === null || typeof raw !== 'object' || Array.isArray(raw)) {
            return null;
        }
        const o = raw as Record<string, Object>;
        const fe = ArenaBilingualParsing.str(o['fullExplanation']);
        return fe.length === 0 ? null : fe;
    }
    private static parseDilemmaSummaryDict(root: Record<string, Object>): DilemmaSummaryBilingualParsed | null {
        const directEn = ArenaBilingualParsing.explanation(root['en']);
        const directZh = ArenaBilingualParsing.explanation(root['zh']);
        if (directEn !== null && directZh !== null) {
            return new DilemmaSummaryBilingualParsed(directEn, directZh);
        }
        const en = ArenaBilingualParsing.explanation(ArenaBilingualParsing.pickLocaleBlock(root, ['en', 'english']));
        const zh = ArenaBilingualParsing.explanation(ArenaBilingualParsing.pickLocaleBlock(root, ['zh', 'chinese', 'cn']));
        if (en !== null && zh !== null) {
            return new DilemmaSummaryBilingualParsed(en, zh);
        }
        const fe = ArenaBilingualParsing.str(root['fullExplanation']);
        if (fe.length > 0) {
            return new DilemmaSummaryBilingualParsed(fe, fe);
        }
        return null;
    }
    static parseRoundtableMessages(sourceText: string, structured: Object | null): RoundtableMessagesBilingualParsed | null {
        if (structured !== null && typeof structured === 'object' && !Array.isArray(structured)) {
            const p = ArenaBilingualParsing.parseRoundtableDict(structured as Record<string, Object>);
            if (p !== null) {
                return p;
            }
        }
        try {
            const root = JSON.parse(sourceText) as Record<string, Object>;
            return ArenaBilingualParsing.parseRoundtableDict(root);
        }
        catch (_e) {
            return null;
        }
    }
    private static messages(raw: Object | undefined): RoundtableMessageSlice[] | null {
        if (raw === undefined || raw === null || typeof raw !== 'object' || Array.isArray(raw)) {
            return null;
        }
        const o = raw as Record<string, Object>;
        const arr = o['messages'];
        if (!Array.isArray(arr)) {
            return null;
        }
        const mapped: RoundtableMessageSlice[] = [];
        for (let i = 0; i < arr.length; i++) {
            const item = arr[i] as Object;
            if (item === null || typeof item !== 'object' || Array.isArray(item)) {
                continue;
            }
            const row = item as Record<string, Object>;
            const speaker = ArenaBilingualParsing.str(row['speaker']);
            const content = ArenaBilingualParsing.str(row['content']);
            if (speaker.length > 0 && content.length > 0) {
                mapped.push(new RoundtableMessageSlice(speaker, content));
            }
        }
        return mapped.length === 0 ? null : mapped;
    }
    private static parseRoundtableDict(root: Record<string, Object>): RoundtableMessagesBilingualParsed | null {
        const directEn = ArenaBilingualParsing.messages(root['en']);
        const directZh = ArenaBilingualParsing.messages(root['zh']);
        if (directEn !== null && directZh !== null) {
            return new RoundtableMessagesBilingualParsed(directEn, directZh);
        }
        const en = ArenaBilingualParsing.messages(ArenaBilingualParsing.pickLocaleBlock(root, ['en', 'english']));
        const zh = ArenaBilingualParsing.messages(ArenaBilingualParsing.pickLocaleBlock(root, ['zh', 'chinese', 'cn']));
        if (en !== null && zh !== null) {
            return new RoundtableMessagesBilingualParsed(en, zh);
        }
        const single = ArenaBilingualParsing.messages(root);
        if (single !== null) {
            return new RoundtableMessagesBilingualParsed(single, single);
        }
        return null;
    }
    static parseDisciplineSummary(sourceText: string, structured: Object | null): DisciplineSummaryBilingualParsed | null {
        if (structured !== null && typeof structured === 'object' && !Array.isArray(structured)) {
            const p = ArenaBilingualParsing.parseDisciplineSummaryDict(structured as Record<string, Object>);
            if (p !== null) {
                return p;
            }
        }
        try {
            const root = JSON.parse(sourceText) as Record<string, Object>;
            return ArenaBilingualParsing.parseDisciplineSummaryDict(root);
        }
        catch (_e) {
            return null;
        }
    }
    private static summary(raw: Object | undefined): string | null {
        if (raw === undefined || raw === null || typeof raw !== 'object' || Array.isArray(raw)) {
            return null;
        }
        const o = raw as Record<string, Object>;
        const s = ArenaBilingualParsing.str(o['summary'] ?? o['reveal'] ?? o['fullExplanation']);
        return s.length === 0 ? null : s;
    }
    private static parseDisciplineSummaryDict(root: Record<string, Object>): DisciplineSummaryBilingualParsed | null {
        const directEn = ArenaBilingualParsing.summary(root['en']);
        const directZh = ArenaBilingualParsing.summary(root['zh']);
        if (directEn !== null && directZh !== null) {
            return new DisciplineSummaryBilingualParsed(directEn, directZh);
        }
        const en = ArenaBilingualParsing.summary(ArenaBilingualParsing.pickLocaleBlock(root, ['en', 'english']));
        const zh = ArenaBilingualParsing.summary(ArenaBilingualParsing.pickLocaleBlock(root, ['zh', 'chinese', 'cn']));
        if (en !== null && zh !== null) {
            return new DisciplineSummaryBilingualParsed(en, zh);
        }
        const single = ArenaBilingualParsing.summary(root);
        if (single !== null) {
            return new DisciplineSummaryBilingualParsed(single, single);
        }
        return null;
    }
    static parseDisciplineDual(sourceText: string, structured: Object | null): DisciplineDualReplyParsed | null {
        if (structured !== null && typeof structured === 'object' && !Array.isArray(structured)) {
            const dict = structured as Record<string, Object>;
            const b = dict['builder'];
            const k = dict['breaker'];
            if (typeof b === 'string' && typeof k === 'string') {
                const builder = b.trim();
                const breaker = k.trim();
                if (builder.length > 0 && breaker.length > 0) {
                    return new DisciplineDualReplyParsed(builder, breaker);
                }
            }
        }
        return ArenaBilingualParsing.parseDisciplineDualMarkers(ArenaBilingualParsing.finalizeStreamSpeech(sourceText));
    }
    static finalizeStreamSpeech(raw: string): string {
        const trimmed = raw.trim();
        if (!trimmed.startsWith('{')) {
            return trimmed;
        }
        try {
            const o = JSON.parse(trimmed) as Record<string, Object>;
            const c = o['content'];
            if (typeof c === 'string' && c.length > 0) {
                return c;
            }
        }
        catch (_e) {
            // fall through
        }
        return trimmed;
    }
    private static parseDisciplineDualMarkers(sourceText: string): DisciplineDualReplyParsed | null {
        const patternBuilder = /^\s*(?:\[Builder\]|【建构者】|建构者[:：])\s*/m;
        const patternBreaker = /^\s*(?:\[Breaker\]|【破坏者】|破坏者[:：])\s*/m;
        const bMatch = patternBuilder.exec(sourceText);
        const kMatch = patternBreaker.exec(sourceText);
        if (bMatch === null || kMatch === null || bMatch.index === undefined || kMatch.index === undefined) {
            return null;
        }
        if (bMatch.index >= kMatch.index) {
            return null;
        }
        const builderStart = bMatch.index + bMatch[0].length;
        const builderText = sourceText.substring(builderStart, kMatch.index).trim();
        const breakerStart = kMatch.index + kMatch[0].length;
        const breakerText = sourceText.substring(breakerStart).trim();
        if (builderText.length === 0 || breakerText.length === 0) {
            return null;
        }
        return new DisciplineDualReplyParsed(builderText, breakerText);
    }
    static buildDebateNoteKey(philosopherId: string, question: string): string {
        return `${philosopherId}|${question.trim()}`;
    }
}
